libtiff/libtiff_ctypes.py

```diff
@@ -829,6 +829,132 @@ def write_plane(arr, tile_arr, width, height, plane_index=0, depth_index=0):
+        this_tile_height = min(num_trows, num_irows - y)
+        this_tile_width = min(num_tcols, num_icols - x)
+
+        def read_plane(tile_plane, plane_index=0, depth_index=0):
+            """ Read one plane from TIFF. The TIFF has more than one plane only if
+            it has more than one sample per pixel, and planar_config == PLANARCONFIG_SEPARATE
+            """
+            # the numpy array should be contigous in memory before calling ReadTile
+            tile_plane = np.ascontiguousarray(tile_plane)
+            # even if the tile is on the edge, and the final size will be smaller,
+            # the size of the array passed to the ReadTile function
+            # must be (num_tcols, num_trows)
+            #
+            # The image has only one depth (ImageDepth == 1), so
+            # the z parameter is not read
+            r = self.ReadTile(tile_plane.ctypes.data, x, y, depth_index, plane_index)
+            if not r:
+                raise ValueError(
+                    "Could not read tile x:%d,y:%dfrom file" % (x, y))
+
+            # check if the tile is on the edge of the image
+            if this_tile_height < num_trows or this_tile_width < num_tcols:
+                # if the tile is on the edge of the image, generate a smaller tile
+                tile_plane = tile_plane[:this_tile_height, :this_tile_width]
+
+            return tile_plane
+
+        if num_idepth == 1:
+            if samples_pp == 1:
+                # the tile plane has always the size of a full tile
+                tile_plane = np.zeros((num_trows, num_tcols), dtype=dtype)
+                # this tile may be smaller than tile_plane
+                tile = read_plane(tile_plane)
+            else:
+                if planar_config == PLANARCONFIG_CONTIG:
+                    # the tile plane has always the size of a full tile
+                    tile_plane = np.empty((num_trows, num_tcols, samples_pp), dtype=dtype)
+                    # this tile may be smaller than tile_plane,
+                    # if the tile is on the edge of the image
+                    tile = read_plane(tile_plane)
+                else:
+                    # the tile plane has always the size of a full tile
+                    tile_plane = np.empty((samples_pp, num_trows, num_tcols), dtype=dtype)
+                    # this tile may be smaller than tile_plane,
+                    # if the tile is on the edge of the image
+                    tile = np.empty((samples_pp, this_tile_width, this_tile_height), dtype=dtype)
+                    for plane_index in xrange(samples_pp):
+                        tile[plane_index] = read_plane(tile[plane_index], plane_index)
+
+        else:
+            if samples_pp > 1:
+                raise NotImplementedError("ImageDepth > 1 and SamplesPerPixel > 1 not implemented")
+
+            # the tile plane has always the size of a full tile
+            tile_plane = np.zeros((samples_pp, num_trows, num_tcols), dtype=dtype)
```

Comment:
Here, it's not samples_pp, but num_idepth, right?
`tile_plane = np.empty((num_idepth, num_trows, num_tcols), dtype=dtype)`